File: lintcode/minimum-path-sum.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/minimum-path-sum/

class Solution:
    """
    @param grid: a list of lists of integers.
    @return: An integer, minimizes the sum of all numbers along its path
    """
    def minPathSum(self, grid):
        # write your code here
        row = len(grid)
        if row == 0:
            return 0
        col = len(grid[0])
        if col == 0:
            return 0

        dp = [[0] * col for i in xrange(row)]
        dp[row-1][col-1] = grid[row-1][col-1]
        for i in xrange(row-2, -1, -1):
            dp[i][col-1] = dp[i+1][col-1] + grid[i][col-1]
        for j in xrange(col-2, -1, -1):
            dp[row-1][j] = dp[row-1][j+1] + grid[row-1][j]

        for i in xrange(row-2, -1, -1):
            for j in xrange(col-2, -1, -1):
                dp[i][j] = min(dp[i+1][j], dp[i][j+1]) + grid[i][j]

        return dp[0][0]

if __name__ == '__main__':
    s = Solution()
    print s.minPathSum([[0]])

File: leetcode-2/minimum-path-sum.cc
// Question: https://leetcode.com/problems/minimum-path-sum/

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        if (grid.size() == 0 || grid[0].size() == 0) {
            return 0;
        }

        int row = grid.size();
        int col = grid[0].size();
        vector<vector<int>> min_sum(row, vector<int>(col, INT_MAX));
        min_sum[0][0] = grid[0][0];
        for (int i = 1; i < row; ++i) {
            min_sum[i][0] = min_sum[i-1][0] + grid[i][0];
        }
        for (int j = 1; j < col; ++j) {
            min_sum[0][j] = min_sum[0][j-1] + grid[0][j];
        }

        for (int i = 1; i < row; ++i) {
            for (int j = 1; j < col; ++j) {
                min_sum[i][j] = min(min_sum[i-1][j], min_sum[i][j-1]) + grid[i][j];
            }
        }
        return min_sum.back().back();
    }
};
